adapters/adhese/adhese.go

```diff
@@ -0,0 +1,274 @@
+			return nil, []error{err, WrapError(fmt.Sprintf("Response %v could not be parsed as Adhese bid.", string(response.Body)))}
+		}
+		bidResponse = convertAdheseBid(adheseBidResponseArray[0])
+	} else {
+		var openRtbBidResponseArray []openrtb_ext.AdheseOpenRtbBid
+		if err := json.Unmarshal(response.Body, &openRtbBidResponseArray); err != nil {
+			return nil, []error{err, WrapError(fmt.Sprintf("Response %v could not be parsed as Adhese OpenRtb Bid.", string(response.Body)))}
+		}
+		bidResponse = convertAdheseOpenRtbBid(openRtbBidResponseArray[0])
+	}
+
+	bidderResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	if len(bidResponse.SeatBid) == 0 {
+		return nil, []error{WrapError("Response resulted in an empty seatBid array.")}
+	}
+
+	var errs []error
+	for _, sb := range bidResponse.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+			bidderResponse.Bids = append(bidderResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: getBidType(bid.AdM),
+			})
+
+		}
+	}
+	return bidderResponse, errs
+}
+
+func convertAdheseBid(adheseBid openrtb_ext.AdheseBid) openrtb.BidResponse {
+	price, _ := strconv.ParseFloat(adheseBid.Extension.Prebid.Cpm.Amount, 64)
+	width, _ := strconv.ParseUint(adheseBid.Width, 10, 64)
+	height, _ := strconv.ParseUint(adheseBid.Height, 10, 64)
+	adheseObj, _ := json.Marshal(openrtb_ext.ExtAdhese{
+		CreativeId:                adheseBid.Id,
+		AdFormat:                  adheseBid.AdFormat,
+		AdType:                    adheseBid.AdType,
+		AdspaceId:                 adheseBid.AdspaceId,
+		DealId:                    adheseBid.OrderId,
+		LibId:                     adheseBid.LibId,
+		OrderProperty:             adheseBid.OrderProperty,
+		Priority:                  adheseBid.Priority,
+		ViewableImpressionCounter: adheseBid.ViewableImpressionCounter,
+	})
+	return openrtb.BidResponse{
+		ID: adheseBid.Id,
+		SeatBid: []openrtb.SeatBid{{
+			Bid: []openrtb.Bid{{
+				ID:     adheseBid.Id,
+				ImpID:  adheseBid.CreativeName,
+				DealID: adheseBid.OrderId,
+				Price:  price,
+				W:      width,
+				H:      height,
+				CID:    adheseBid.OrderId,
+				CrID:   adheseBid.Id,
+				BURL:   adheseBid.Tracker,
+				AdM:    getAdMarkup(adheseBid),
+				Ext:    adheseObj,
+			}},
+			Seat: "",
+		}},
+		BidID: adheseBid.OrderId,
+		Cur:   adheseBid.Extension.Prebid.Cpm.Currency,
+	}
+}
+
+func convertAdheseOpenRtbBid(adheseBid openrtb_ext.AdheseOpenRtbBid) openrtb.BidResponse {
+	price, _ := strconv.ParseFloat(adheseBid.Extension.Prebid.Cpm.Amount, 64)
+	width, _ := strconv.ParseUint(adheseBid.Width, 10, 64)
+	height, _ := strconv.ParseUint(adheseBid.Height, 10, 64)
+	var response openrtb.BidResponse = adheseBid.OriginData
+	response.ID = adheseBid.Origin
+	if adheseBid.OriginInstance != "" {
+		response.ID = response.ID + "-" + adheseBid.OriginInstance
+	}
+	if len(response.SeatBid) > 0 && len(response.SeatBid[0].Bid) > 0 {
```

Comment:
This if looks suspicious to me - the origin data is kind of optional, so it's possible that it does not exist at all and we still have a valid bid. 
I think should create a new SeatBid with all the price, width, etc. and merge it somehow with the (optional) structure contained in originData.

Also `adheseBid.Extension.Prebid` is kind of optional - can be absent if preprocessor isn't configured for prebid. Maybe we should check for it and fail gracefully (with a warning?) in such case?

Reply:
Why does it need to fail if `adheseBid.Extension.Prebid` is missing while optional? Can't we just assume the default currency USD?

For the first part - are you saying that `adheseBid.OriginData` might be missing or null?

Reply:
1. You can assume the currency is USD, but what will you assume about the price which would be missing too?
2. I looked into the code and it looks like originData is always added... so my argument is not so valid... although I can imagine an implementation of a market that wouldn't add originData field at all - howevernot the case now

Reply:
For 2; Then I prefere to leave it as is, because merging objects in go requires some external libraries or tedious programming